lib/plugins/aws/package/lib/resolveIamRoles.js

```diff
@@ -0,0 +1,69 @@
+'use strict';
+
+const _ = require('lodash');
+
+module.exports = {
+  resolveIamRoles() {
+    const resource = this.serverless.service.provider.compiledCloudFormationTemplate.Resources[
+      this.provider.naming.getRoleLogicalId()
+    ].Properties;
+
+    // Register all awsProvider.iamConfig.principals and functions[].iamConfig.principals
+    // on IamRoleLambdaExecution resource (ensuring no duplicates)
+    const CFprincipals = resource.AssumeRolePolicyDocument.Statement[0].Principal.Service;
+
+    const principals = new Set(this.serverless.service.provider.iamConfig.principals);
+
+    this.serverless.service.getAllFunctions().forEach(functionName => {
+      const functionObject = this.serverless.service.getFunction(functionName);
+
+      functionObject.iamConfig.principals.forEach(principal => {
+        principals.add(principal);
+      });
+    });
+
+    principals.forEach(principal => {
+      CFprincipals.push(principal);
+    });
+
+    // add all awsProvider.iamConfig.managedPolicies and functions[].iamConfig.managedPolicies
+    // and functions[].iamConfig.managedPolicies to IamRoleLambdaExecution resource (ensuring no duplicates)
+
+    let managedPolicies = this.serverless.service.provider.iamConfig.managedPolicies;
+
+    this.serverless.service.getAllFunctions().forEach(functionName => {
+      const functionObject = this.serverless.service.getFunction(functionName);
+
+      managedPolicies = managedPolicies.concat(functionObject.iamConfig.managedPolicies);
+    });
+
+    managedPolicies = _.uniqWith(managedPolicies, _.isEqual);
+
+    if (managedPolicies.length > 0) {
+      if (!Array.isArray(resource.ManagedPolicyArns)) {
+        resource.ManagedPolicyArns = [];
+      }
+      resource.ManagedPolicyArns = resource.ManagedPolicyArns.concat(managedPolicies);
+    }
+
+    // add all awsProvider.iamConfig.policyStatements and functions[].iamConfig.policyStatements to IamRoleLambdaExecution resource (ensuring no duplicates)
+
+    let policyStatements = this.serverless.service.provider.iamConfig.policyStatements;
```

Comment:
I think we don't need to filter statements with empty `Resources` section since we're adding all statements from `iamRoleProperties.Policies[0].PolicyDocument.Statement` to the temporary array and then filtering out any duplicates. Take a look at [lines 87-91](https://github.com/serverless/serverless/pull/8511/commits/34b66dd6135e2155500a968a376278ac36815a7b#diff-3dd0bb69248845944afeb1f1f276173560186609a582512d4d9d6bf79977e9abR87-R91) in latest commit and let me know!

Reply:
Will filtering the duplicates, filter statements as that one: https://github.com/serverless/serverless/blob/d8527d8b57e7e5f0b94ba704d9f53adb34298d99/lib/plugins/aws/package/lib/iam-role-lambda-execution-template.json#L22-L26  ?

My point was, to not have such statements at all

Reply:
Oh I see. In that case, I will add some code to take those statements out completely.